File: bindings/dotnet/src/Maplibre.Native/Internal/Struct/ResourceStructs.cs
using System.Runtime.InteropServices;
using Maplibre.Native.Internal.C;
using Maplibre.Native.Internal.Memory;
using Maplibre.Native.Resource;

namespace Maplibre.Native.Internal.Struct;

internal static unsafe class ResourceStructs
{
    internal static ResourceRequest ReadRequest(mln_resource_request* request)
    {
        if (request is null)
        {
            throw new ArgumentNullException(nameof(request));
        }

        return new ResourceRequest(
            (ResourceKind)request->kind,
            CopyUtf8(request->url),
            (ResourceLoadingMethod)request->loading_method,
            (ResourcePriority)request->priority,
            (ResourceUsage)request->usage,
            (ResourceStoragePolicy)request->storage_policy,
            request->has_range != 0
                ? new ByteRange(request->range_start, request->range_end)
                : null,
            request->has_prior_modified != 0
                ? DateTimeOffset.FromUnixTimeMilliseconds(request->prior_modified_unix_ms)
                : null,
            request->has_prior_expires != 0
                ? DateTimeOffset.FromUnixTimeMilliseconds(request->prior_expires_unix_ms)
                : null,
            CopyNullableUtf8(request->prior_etag),
            request->prior_data_size == 0 ? null : request->prior_data_size,
            CopyBytes(request->prior_data, request->prior_data_size)
        );
    }

    private static string CopyUtf8(sbyte* value) =>
        value is null ? string.Empty : Marshal.PtrToStringUTF8((nint)value) ?? string.Empty;

    private static string? CopyNullableUtf8(sbyte* value) =>
        value is null ? null : Marshal.PtrToStringUTF8((nint)value);

    private static byte[]? CopyBytes(byte* value, nuint byteLength)
    {
        if (value is null || byteLength == 0)
        {
            return null;
        }

        var bytes = new byte[checked((int)byteLength)];
        Marshal.Copy((nint)value, bytes, 0, bytes.Length);
        return bytes;
    }
}

internal sealed unsafe class NativeResourceResponse : IDisposable
{
    private readonly nint bytes;
    private readonly NativeUtf8String errorMessage;
    private readonly NativeUtf8String etag;

    private NativeResourceResponse(
        mln_resource_response value,
        nint bytes,
        NativeUtf8String errorMessage,
        NativeUtf8String etag
    )
    {
        Value = value;
        this.bytes = bytes;
        this.errorMessage = errorMessage;
        this.etag = etag;
    }

    internal mln_resource_response Value { get; }

    internal static NativeResourceResponse From(ResourceResponse response)
    {
        ArgumentNullException.ThrowIfNull(response);

        nint bytesPointer = 0;
        if (response.Bytes.Length > 0)
        {
            bytesPointer = (nint)NativeMemory.Alloc((nuint)response.Bytes.Length);
            Marshal.Copy(response.Bytes, 0, bytesPointer, response.Bytes.Length);
        }

        try
        {
            var errorMessage = NativeUtf8String.FromNullableString(
                response.ErrorMessage,
                nameof(response.ErrorMessage)
            );
            try
            {
                var etag = NativeUtf8String.FromNullableString(
                    response.Etag,
                    nameof(response.Etag)
                );
                var value = new mln_resource_response
                {
                    size = (uint)sizeof(mln_resource_response),
                    status = (uint)response.Status,
                    error_reason = (uint)response.ErrorReason,
                    bytes = (byte*)bytesPointer,
                    byte_count = (nuint)response.Bytes.Length,
                    error_message = errorMessage.Pointer,
                    must_revalidate = response.MustRevalidate ? (byte)1 : (byte)0,
                    has_modified = response.Modified.HasValue ? (byte)1 : (byte)0,
                    modified_unix_ms = response.Modified?.ToUnixTimeMilliseconds() ?? 0,
                    has_expires = response.Expires.HasValue ? (byte)1 : (byte)0,
                    expires_unix_ms = response.Expires?.ToUnixTimeMilliseconds() ?? 0,
                    etag = etag.Pointer,
                    has_retry_after = response.RetryAfter.HasValue ? (byte)1 : (byte)0,
                    retry_after_unix_ms = response.RetryAfter?.ToUnixTimeMilliseconds() ?? 0,
                };
                return new NativeResourceResponse(value, bytesPointer, errorMessage, etag);
            }
            catch
            {
                errorMessage.Dispose();
                throw;
            }
        }
        catch
        {
            if (bytesPointer != 0)
            {
                NativeMemory.Free((void*)bytesPointer);
            }
            throw;
        }
    }

    public void Dispose()
    {
        errorMessage.Dispose();
        etag.Dispose();
        if (bytes != 0)
        {
            NativeMemory.Free((void*)bytes);
        }
    }
}

File: bindings/kotlin/src/nativeMain/kotlin/org/maplibre/nativeffi/internal/struct/ResourceStructs.kt
package org.maplibre.nativeffi.internal.struct

import kotlinx.cinterop.ByteVar
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.MemScope
import kotlinx.cinterop.alloc
import kotlinx.cinterop.ptr
import kotlinx.cinterop.readBytes
import kotlinx.cinterop.toCValues
import org.maplibre.nativeffi.internal.c.mln_resource_request
import org.maplibre.nativeffi.internal.c.mln_resource_response
import org.maplibre.nativeffi.internal.memory.MemoryUtil
import org.maplibre.nativeffi.internal.status.Status
import org.maplibre.nativeffi.resource.ResourceKind
import org.maplibre.nativeffi.resource.ResourceLoadingMethod
import org.maplibre.nativeffi.resource.ResourcePriority
import org.maplibre.nativeffi.resource.ResourceRequest
import org.maplibre.nativeffi.resource.ResourceResponse
import org.maplibre.nativeffi.resource.ResourceStoragePolicy
import org.maplibre.nativeffi.resource.ResourceUsage

/** Internal materializers and readers for resource callback structs. */
@OptIn(ExperimentalForeignApi::class, ExperimentalUnsignedTypes::class)
internal object ResourceStructs {
  fun resourceRequest(value: mln_resource_request): ResourceRequest {
    val rawKind = value.kind
    val rawLoadingMethod = value.loading_method
    val rawPriority = value.priority
    val rawUsage = value.usage
    val rawStoragePolicy = value.storage_policy
    return ResourceRequest(
      MemoryUtil.copyCString(value.url),
      ResourceKind.fromNative(rawKind),
      ResourceLoadingMethod.fromNative(rawLoadingMethod),
      ResourcePriority.fromNative(rawPriority),
      ResourceUsage.fromNative(rawUsage),
      ResourceStoragePolicy.fromNative(rawStoragePolicy),
      if (value.has_range)
        ResourceRequest.ByteRange(
          uint64BitsToLong(value.range_start),
          uint64BitsToLong(value.range_end),
        )
      else null,
      if (value.has_prior_modified) value.prior_modified_unix_ms else null,
      if (value.has_prior_expires) value.prior_expires_unix_ms else null,
      optionalCString(value.prior_etag),
      value.prior_data?.readBytes(checkedInt(value.prior_data_size, "prior data size"))
        ?: ByteArray(0),
    )
  }

  fun resourceResponse(value: ResourceResponse, scope: MemScope): CPointer<mln_resource_response> {
    val bytes = value.bytes
    val native = scope.alloc<mln_resource_response>()
    native.size = kotlinx.cinterop.sizeOf<mln_resource_response>().toUInt()
    native.status = value.status.nativeValue.toUInt()
    if (!value.errorReason.isKnown) {
      throw Status.invalidArgument(
        "Unknown resource error reason cannot be used as input: ${value.errorReason.nativeValue}"
      )
    }
    native.error_reason = value.errorReason.nativeValue.toUInt()
    if (bytes.isNotEmpty()) {
      native.bytes = bytes.toUByteArray().toCValues().getPointer(scope)
      native.byte_count = bytes.size.toULong()
    }
    value.errorMessage?.let {
      native.error_message = resourceResponseCString(scope, it, "error message")
    }
    native.must_revalidate = value.mustRevalidate
    value.modifiedUnixMs?.let {
      native.has_modified = true
      native.modified_unix_ms = it
    }
    value.expiresUnixMs?.let {
      native.has_expires = true
      native.expires_unix_ms = it
    }
    value.etag?.let { native.etag = resourceResponseCString(scope, it, "ETag") }
    value.retryAfterUnixMs?.let {
      native.has_retry_after = true
      native.retry_after_unix_ms = it
    }
    return native.ptr
  }

  private fun uint64BitsToLong(value: ULong): Long = value.toLong()

  private fun checkedInt(value: ULong, name: String): Int {
    require(value <= Int.MAX_VALUE.toULong()) { "$name exceeds Int.MAX_VALUE" }
    return value.toInt()
  }

  private fun resourceResponseCString(
    scope: MemScope,
    value: String,
    description: String,
  ): CPointer<ByteVar> {
    if ('\u0000' in value) {
      throw Status.invalidArgument("$description contains embedded NUL")
    }
    return MemoryUtil.cString(scope, value)
  }

  private fun optionalCString(value: CPointer<ByteVar>?): String? =
    if (value == null) null else MemoryUtil.copyCString(value)
}
